### PR TITLE
MMS testing

### PR DESCRIPTION
**Pull request by jhdark**
## Proposed changes

MMS testing! This PR starts to build up an MMS testing suite for vigorous verification testing.  @RemDelaporteMathurin See what you think so far. I've adapted the tests in [fenicsx tutorial](https://jsdokken.com/dolfinx-tutorial/chapter4/convergence.html).

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: test/test_mms.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    V = fem.FunctionSpace(test_mesh_1d.mesh, elements)
+    exact = fem.Function(V)
+    exact.interpolate(u_numpy)
+    exact_xdmf = XDMFFile(MPI.COMM_WORLD, "results/mms/exact_solution.xdmf", "w")
+    exact_xdmf.write_mesh(test_mesh_1d.mesh)
+    exact_xdmf.write_function(exact)
+
+    D_0 = 1
+    E_D = 0
+    D = D_0 * ufl.exp(-E_D / (F.k_B * 1))
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:63]: ⚠️ avoid unity parameters whenever possible

This hides away potential bugs. Imagine that in FESTIM we mistakenly coded:

$$ \nabla \cdot (\nabla c) = 0$$

instead of 

$$ \nabla \cdot (D\nabla c) = 0$$

Then this test wouldn't catch this bug.

File: test/test_mms.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    vol = F.VolumeSubdomain1D(id=1, borders=[0, 1], material=my_mat)
+    left = F.SurfaceSubdomain1D(id=1, x=0)
+    right = F.SurfaceSubdomain1D(id=2, x=1)
+
+    my_model.subdomains = [vol, left, right]
+
+    A = F.Species("A")
+    my_model.species = [A]
+
+    my_model.temperature = 1
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:77]: Here you hard coded the temperature as one (see comment about unity).
You also hard coded it above in the expression of the diffusivity.

This is error prone as a change in one place wouldn't be carried over in the rest of the code.

I would create a `T` variable

File: test/test_mms.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    my_model.boundary_conditions = [
+        F.DirichletBC(subdomain=left, value=u_ufl, species=A),
+        F.DirichletBC(subdomain=right, value=u_ufl, species=A),
+    ]
+
+    f_expr = lambda x: D * 4 * np.pi**2 * np.sin(2 * np.pi * x[0])
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:84]: So you derived this source term by hand, correct?

This is fine for simple cases, but ultimately we want to test complex cases like space varying temperature, etc.

Therefore, we need to look at symbolic differenciation with sympy.

You can "lambdify" sympy expression very easily.

File: test/test_mms.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    my_model.mesh = test_mesh_1d
+    my_mat = F.Material(name="mat", D_0=D_0, E_D=E_D)
+    vol = F.VolumeSubdomain1D(id=1, borders=[0, 1], material=my_mat)
+    left = F.SurfaceSubdomain1D(id=1, x=0)
+    right = F.SurfaceSubdomain1D(id=2, x=1)
+
+    my_model.subdomains = [vol, left, right]
+
+    A = F.Species("A")
+    my_model.species = [A]
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:75]: Why not "H" to avoid confusion?

File: test/test_mms.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    my_model.mesh = test_mesh_1d
+    my_mat = F.Material(name="mat", D_0=D_0, E_D=E_D)
+    vol = F.VolumeSubdomain1D(id=1, borders=[0, 1], material=my_mat)
+    left = F.SurfaceSubdomain1D(id=2, x=0)
+    right = F.SurfaceSubdomain1D(id=3, x=1)
+    my_model.subdomains = [vol, left, right]
+
+    A = F.Species("A")
+    B = F.Species("B", mobile=False)
+    traps = F.ImplicitSpecies(n=n_trap, others=[B])
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:151]: again, I would choose variable names that are more explicit

**Comment by jhdark** [comment]
@RemDelaporteMathurin here are basic examples of the types of MMS tests we could do. But perhaps we can discuss further in the general channel on Slack for what should be covered with MMS testing

File: test/test_mms.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    L2_error_mobile = error_L2(mobile_computed, mobile_analytical_np)
+    L2_error_trapped = error_L2(trapped_computed, trapped_analytical_np)
+
+    assert L2_error_mobile < 2e-07
+    assert L2_error_trapped < 1e-07
+
+
+def test_MMS_transient():
+    """
+    MMS test with one mobile species in 0.1s transient
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:209]: I would precise here that you only check the last timestep

File: test/test_mms.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    init_value = lambda x: 1 + ufl.sin(2 * ufl.pi * x[0])
+    my_model.initial_conditions = [F.InitialCondition(value=init_value, species=H)]
+
+    my_model.settings = F.Settings(atol=1e-10, rtol=1e-10, final_time=final_time)
+    my_model.settings.stepsize = final_time / 20
+
+    my_model.initialise()
+
+    f = lambda x, t: (
+        (H_analytical_ufl(x, t) - H.prev_solution) / my_model.dt
```
Comment (RemDelaporteMathurin) [inline test/test_mms.py:263]: No you shouldn't have this in MMS.
If your manufactured solution is:

$$c =  1 + \sin{2 \pi \ x} + t^2$$

then

$$ \frac{\partial c}{\partial t} = 2 \ t $$


The form c - c_n / delta_t is not an exact solution, it's an approximation of dc/dt !!

**Comment by RemDelaporteMathurin** [comment]
We had a discussion with @jhdark and agreed that while this may not be the most generic way to do MMS, we should go with this for now and think later of a way to refactor these tests as well as define a test matrix for all the cases.

File: test/test_mms.py
Comment (RemDelaporteMathurin) [inline test/test_mms.py:1]: what's the difference between test_mms and the system subfolder?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]

File: test/system/test_1_mobile_species.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+    my_model.run()
+
+    H_computed = H.post_processing_solution
+
+    L2_error = error_L2(H_computed, H_analytical_np)
+
+    assert L2_error < 1e-2
+
+
+def test_1_mobile_MMS_multivolume():
```
[the remark this answers is not in the document]
Comment (jhdark) [reply test/system/test_1_mobile_species.py:305]: ok we can leave it out

File: test/system/test_1_mobile_species.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+    H_analytical_np = u_exact(np)
+
+    elements = ufl.FiniteElement("CG", test_mesh_3d.ufl_cell(), 1)
+    V = fem.FunctionSpace(test_mesh_3d, elements)
+    T = fem.Function(V)
+
+    D_0 = 1
+    E_D = 0.1
+    T_expr = lambda x: 500 + 100 * x[0]
+    T.interpolate(T_expr)
```
[the remark this answers is not in the document]
Comment (jhdark) [reply test/system/test_1_mobile_species.py:239]: Unfortunately, this is needed for the source term. We could get around this once we have a heat transfer module to extract the temperature function that way.

File: test/system/test_1_mobile_species.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+    D = D_0 * ufl.exp(-E_D / (F.k_B * T))
+
+    my_model = F.HydrogenTransportProblem()
+    my_model.mesh = F.Mesh(mesh=test_mesh_2d)
+
+    # create facet meshtags
+    boundaries = [
+        (1, lambda x: np.isclose(x[0], 0)),
+        (2, lambda x: np.isclose(x[0], 1)),
+    ]
```
[the remark this answers is not in the document]
Comment (jhdark) [reply test/system/test_1_mobile_species.py:166]: This is true, we will need a more generic implementation for when we expand this testing.

File: test/system/test_1_mobile_species.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply test/system/test_1_mobile_species.py:239]: Needing a full heat transfer problem class just for this sounds a bit overkill doesnt it?

I guess we will merge this but it will require a big rewrite at some point to homogeneise and make it generic...

File: festim/hydrogen_transport_problem.py [hunk 11]
```diff
@@ -9,6 +10,8 @@
 import tqdm.autonotebook
 import festim as F
 
+from dolfinx.mesh import meshtags
+
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:14]:
```suggestion
```

already imported

File: festim/exports/surface_quantity.py
Comment (RemDelaporteMathurin) [inline festim/exports/surface_quantity.py:1]: I don't understand why this change is in this PR as it already exists in _fenicsx_ maybe rebase?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]

**Comment by jhdark** [comment]
I rebased from fenicsx, but theres still, if not more, changes that are already in that branch? Not sure if I should merge or not?

**Comment by jhdark** [comment]
Remaking PR due to bugs from not detecting changes to fenicsx branch
